content/blog/nats-operator-and-service-accounts.md

```diff
@@ -0,0 +1,337 @@
+natscluster.nats.io/example-nats created
+serviceaccount/nats-admin-user created
+serviceaccount/nats-user created
+nnatsservicerole.nats.io/nats-user unchanged
+natsservicerole.nats.io/nats-admin-user unchanged
+pod/nats-user-pod created
+pod/nats-admin-user-pod created
+```
+
+As it can be seen from the output, the command above has created a couple
+of service accounts, one for a `nats-user` and another one for a
+`nats-admin-user`, as well as a couple of `natsservicerole` objects with
+the same names.  The `NatsServiceRole` is a new
+`CustomResourceDefinition` that the NATS Operator creates to represent
+a `ServiceAccount` that will be used by a NATS cluster.
+
+For example, in the case of the `nats-user`, a `NatsServiceRole` is
+defined as below so that the `example-nats` cluster that was created by the
+operator sets PUB/SUB permissions for that account:
+
+```yaml
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: nats-user
+---
+apiVersion: nats.io/v1alpha2
+kind: NatsServiceRole
+metadata:
+  name: nats-user
+  namespace: nats-io
+
+  # Specifies which NATS cluster will be using this role.
+  labels:
+    nats_cluster: example-nats
+spec:
+  permissions:
+    publish: ["foo.*", "foo.bar.quux"]
+    subscribe: ["foo.bar", "greetings", "hello.world"]
+```
+
+The NATS Operator will then detect that there is a mapping between a
+NatsServiceRole and a ServiceAccount and then issue a new token where
+the audience is set to be the operated NATS cluster service and storing
+it into a secret:
+
+```sh
+kubectl -n nats-io get secret nats-user-example-nats-bound-token  -o yaml
+apiVersion: v1
+data:
+  token: ZXlKaGJHY2lPaUpTVXp...
+kind: Secret
+metadata:
+  creationTimestamp: 2018-07-20T00:48:25Z
+  labels:
+    app: nats
+    nats_cluster: example-nats
+  name: nats-user-example-nats-bound-token
+  namespace: nats-io
+```
+
+Now a Pod in order to use the `nats-user` account, can then mount the secret
+as a volume to be able to use that token to authenticate:
+
+```yaml
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nats-user-pod
+  labels:
+    nats_cluster: example-nats
+spec:
+  volumes:
+    - name: "token"
+      projected:
+        sources:
+        - secret:
+            name: "nats-user-example-nats-bound-token"
+            items:
+              - key: token
+                path: "token"
+  restartPolicy: Never
+  containers:
+    - name: nats-ops
+      command: ["/bin/sh"]
+      image: "wallyqs/nats-ops:latest"
+      tty: true
+      stdin: true
+      stdinOnce: true
+
+      # Service Account Token is mounted via projected volume.
+      volumeMounts:
+      - name: "token"
+        mountPath: "/var/run/secrets/nats.io"
+        readOnly: true
+```
+
+Using `kubectl attach` we can run a few commands in that container to confirm
+that the permissions have been applied to the cluster:
+
```

Comment:
Using `kubectl attach` we can run a few commands in that container to confirm the permissions have been applied to the cluster: